.github/workflows/ci.yaml

```diff
@@ -9,19 +9,104 @@ on:
     types: [ published ]
 
 jobs:
+
+  lint:
+    name: Linter
+    runs-on: ubuntu-latest
+    timeout-minutes: 5
+    outputs:
+      version: ${{ steps.version.outputs.version }}
+    steps:
+      - name: Checkout commit
+        uses: actions/checkout@v2
+      - name: Setup Python 3.7
+        uses: actions/setup-python@v2
+        with:
+          python-version: "3.7"
+      - name: Cache PyPI
+        uses: actions/cache@v2
+        with:
+          key: pip-lint-${{ hashFiles('requirements/*.txt') }}
+          path: ~/.cache/pip
+          restore-keys: |
+              pip-lint-
+      - name: Install dependencies
+        run: |
+          python -m pip install twine wheel
+          make setup
+      - name: Run linters
+        run: |
+          make lint
+      - name: Save the package version
+        id: version
+        run: |
+          echo "::set-output name=version::$(python setup.py --version)"
+      - name: Show version
+        run: |
+            echo ${{ steps.version.outputs.version }}
+
+  pretest:
+    name: Prepare for e2e tests
+    needs: [lint]
+    if: github.event_name != 'release'
+    runs-on: ubuntu-latest
+    env:
+      NEURO_STAGING_URL: ${{ secrets.NEURO_STAGING_URL }}
+      NEURO_TOKEN: ${{ secrets.NEURO_TOKEN }}
+      NEURO_CLUSTER: neuro-compute
+      DOCKER_SERVER: docker.io
+      DOCKER_USERNAME: ${{ secrets.DOCKER_USERNAME }}
+      DOCKER_PASSWORD: ${{ secrets.DOCKER_PASSWORD }}
+      # Note: ${{ github.sha }} not working, see https://github.com/actions/checkout/issues/299
+      SHA: ${{ github.event.pull_request.head.sha }}
+    steps:
+      - name: Checkout commit
+        uses: actions/checkout@v2
+      - name: Setup Python 3.7
+        uses: actions/setup-python@v2
+        with:
+          python-version: "3.7"
+      - name: Install and configure neuro
+        run: |
+          python -m pip install neuromation
+          neuro config login-with-token ${{ env.NEURO_TOKEN }} ${{ env.NEURO_STAGING_URL }}
+          neuro config switch-cluster ${{ env.NEURO_CLUSTER }}
+          neuro config show
+      - name: Configure docker
+        run: |
+          docker login $DOCKER_SERVER \
+            --username $DOCKER_USERNAME --password $DOCKER_PASSWORD
+          neuro config docker
+      - name: Build test image
+        run: |
+            export REPO_URL=https://github.com/neuro-inc/neuro-extras.git
+            export PACKAGE="git+$REPO_URL@$SHA"
+            docker build -t neuromation/neuro-extras:$SHA \
+              --build-arg NEURO_EXTRAS_PACKAGE=$PACKAGE .
+      - name: Push test image
+        run: |
+            neuro push neuromation/neuro-extras:$SHA image:neuromation/neuro-extras:$SHA
+
   test:
-    name: Run tests
+    name: Run e2e tests
+    needs: [pretest]
+    if: github.event_name != 'release'
```

Comment:
Hint: if `test` depends on `pretest` *and* `pretest` has condition `github.event_name != 'release'` --- there is no need to repeat it here. If `pretest` is skipped -- the `test` job is skipped as dependency.

Reply:
I know. I intentionally left it for clarity